### PR TITLE
Fixed compilation with giflib version 5

### PR DESCRIPTION
**Pull request by Mailaender**
Based on https://github.com/pld-linux/libgdiplus/blob/master/libgdiplus-giflib.patch

### CODE REVIEW

**Comment by Mailaender** [comment]
Submitted at https://build.opensuse.org/request/show/295390

File: src/gifcodec.c [hunk 1]
```diff
@@ -583,8 +579,11 @@ gdip_load_gif_image (void *stream, GpImage **image, BOOL from_file)
 	}
 
 	FreeExtensionMono(&global_extensions);
+#if (GIFLIB_MAJOR > 5) || ((GIFLIB_MAJOR == 5) && (GIFLIB_MINOR >= 1))
 	DGifCloseFile (gif);
-
+#else
+	DGifCloseFile (gif);
```
Comment (akoeplinger) [inline src/gifcodec.c]: Both the if and the else do the same thing?

File: src/gifcodec.c [hunk 2]
```diff
@@ -599,7 +598,11 @@ gdip_load_gif_image (void *stream, GpImage **image, BOOL from_file)
 
 	if (gif != NULL) {
 		FreeExtensionMono (&global_extensions);
+#if (GIFLIB_MAJOR > 5) || ((GIFLIB_MAJOR == 5) && (GIFLIB_MINOR >= 1))
 		DGifCloseFile (gif);
+#else
+		DGifCloseFile (gif);
```
Comment (akoeplinger) [inline src/gifcodec.c]: same here. I think you're missing the `, NULL` as in the original patch.

**Comment by Mailaender** [comment]
Updated. https://build.opensuse.org/request/show/295392

**Comment by akoeplinger** [comment]
Thanks. I don't have the authority to decide whether this is OK to be merged or not, can you reach out on the mailing list?

**Comment by Mailaender** [comment]
Announced the fix at https://bugzilla.xamarin.com/show_bug.cgi?id=5859

File: src/gifcodec.c [hunk 3]
```diff
@@ -306,9 +302,9 @@ gdip_load_gif_image (void *stream, GpImage **image, BOOL from_file)
 	loop_counter = FALSE;
 
 	if (from_file) {
-		gif = DGifOpen(stream, &gdip_gif_fileinputfunc);
+		gif = DGifOpen(stream, &gdip_gif_fileinputfunc, NULL);
 	} else {
-		gif = DGifOpen (stream, &gdip_gif_inputfunc);
+		gif = DGifOpen (stream, &gdip_gif_inputfunc, NULL);
```
Comment (akoeplinger) [inline src/gifcodec.c]: these need the GIFLIB_MAJOR guards as well, or it will fail to compile on earlier giflibs.

File: src/gifcodec.c [hunk 4]
```diff
@@ -663,9 +666,9 @@ gdip_save_gif_image (void *stream, GpImage *image, BOOL from_file)
 	}
 
 	if (from_file) {
-		fp = EGifOpenFileName (stream, 0);
+		fp = EGifOpenFileName (stream, 0, NULL);
 	} else {
-		fp = EGifOpen (stream, gdip_gif_outputfunc);
+		fp = EGifOpen (stream, gdip_gif_outputfunc, NULL);
```
Comment (akoeplinger) [inline src/gifcodec.c]: same here

**Comment by akoeplinger** [comment]
I just tried this out (with the added guards), but the "make check" tests fail with this patch

**Comment by akoeplinger** [comment]
I added the additional guards here: https://github.com/akoeplinger/libgdiplus/commit/85fb4b4982b8c1f6d93e15982fea64e4e25efa11, feel free to pull that commit in. The libgdiplus tests run fine with it, but it'd be good to run the System.Drawing testsuite against it.

**Comment by Mailaender** [comment]
Updated. Also added unit checks to https://build.opensuse.org/request/show/295418 however there is currently no current SUSE version that still runs giflib 4 so this remains untested from my side.

**Comment by akoeplinger** [comment]
Thanks, looks good :+1: I tested on Ubuntu 14.04, which is still on giflib 4 and seems fine.

**Comment by directhex** [comment]
build

**Comment by directhex** [comment]
https://jenkins.mono-project.com/job/test-libgdiplus-pull-request/7/

**Comment by Mailaender** [comment]
You should probably close https://bugzilla.xamarin.com/show_bug.cgi?id=5859 then.

**Pull request merged**